### PR TITLE
chore(artifacts): rewrite S3 reference artifact tests with `moto` and move to unit tests

### PR DESCRIPTION
**Pull request by tonyyli-wandb**
## What

Replace the hand-rolled `mock_boto` fake used by the S3 *reference* tests with [`moto`](https://github.com/getmoto/moto)'s in-memory S3, and relocate those tests from the system-test suite into a new unit module `tests/unit_tests/test_artifacts/test_references_s3.py` (matching the Azure/GCS pattern from the earlier PRs in this stack).

## Why

- The bespoke `mock_boto` poked canned values straight into the handler — a fixed `e_tag` of `"1234567890abcde"`, a `version_id` of `"1"`, and a `.limit()` that **ignored its argument**. The tests therefore asserted fixture constants, not real handler behavior.
- `moto` is a widely used -- arguably currently standard -- library for mocking AWS python API calls in tests.  It invokes actual `boto3` code paths, so digests now align more closely with the true object ETags and version IDs that S3 would actually assign (normally captured from the `put_object` response).
- The migrated tests never actually talk to the `wandb` backend, so running them under `system_tests` incurs unnecessary overhead (test container setup and a shard slot) for no reason. **They belong in unit tests, where they should run faster** (both locally and in CI).

## Notes

- **`mock_boto` stays** in the system module — the two remaining *backend-dependent* S3 tests (`test_s3_storage_handler_load_path_missing_reference*`) still use it. This PR only removes the backend-free reference tests.
- **Refs are now realistic.** Directory/path cases assert the full keys moto returns (e.g. `s3://my-bucket/my_dir/my_object.pb`) instead of the old fake's truncated `s3://my-bucket/my_dir`.
- **`test_add_s3_max_objects` is `xfail` (strict).** Against a real client, `S3Handler.store_path` fetches at most `max_objects` via `.limit()`, so the `len(entries) > max_objects` guard can never fire — the old mock only "passed" because its `.limit()` ignored the cap and always returned 2 objects. **A followup PR in this stack may fixes the handler to actually enforce the cap and turn this back into a real, meaningful assertion.**
- **The top-level `artifact.digest == "..."` assertions are dropped**: that hash now derives from the real (random) version IDs, so it isn't a stable constant. The manifest-contents assertion is the meaningful check and is kept.
- Adds `moto[s3]` as a dev dependency (to `requirements/requirements_test.txt`) and regenerates the six dev lock files (additions only — no existing pins changed).

## Verification

`tests/unit_tests/test_artifacts/test_references_s3.py` → **8 passed, 1 xfailed** against real `moto` + `boto3`; `ruff check` / `ruff format --check` clean on both touched test files.

Stacked on #12078 (GCS relocation); base will retarget to `main` once that merges.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PiFnDX4duBZbesmcYMrf88

---
_Generated by [Claude Code](https://claude.ai/code/session_01PiFnDX4duBZbesmcYMrf88)_

### CODE REVIEW

**Comment by tonyyli-wandb** [comment]
* **#12091** <a href="https://app.graphite.com/github/pr/wandb/wandb/12091?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#12090** <a href="https://app.graphite.com/github/pr/wandb/wandb/12090?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/wandb/wandb/12090?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12078** <a href="https://app.graphite.com/github/pr/wandb/wandb/12078?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#12077** <a href="https://app.graphite.com/github/pr/wandb/wandb/12077?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: tests/unit_tests/test_artifacts/test_references_s3.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    return hashlib.md5(mock_s3_content).hexdigest()
+
+
+@fixture
+def artifact() -> Artifact:
+    return Artifact(type="dataset", name="data-artifact")
+
+
+@fixture
+def mock_s3(monkeypatch: pytest.MonkeyPatch) -> Iterator[BaseClient]:
```
Comment (pingleiwandb) [inline tests/unit_tests/test_artifacts/test_references_s3.py]: qq: do we have cases that we test with real s3 bucket? I know gcs reference actually have test on real public gcs bucket and has to use get_blob instead of list_blob when adding files...

Also does moto supports directory bucket https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-express-differences.html#s3-express-specifications (similar to GCS's folder bucket)? Though I think we may not handle/run into issue with s3 like we did with GCS @amusipatla-wandb

Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: Which operations would we need?

IIRC moto is pretty transparent about documenting the operations they support/don't support, I think here: https://github.com/getmoto/moto/blob/master/IMPLEMENTATION_COVERAGE.md

(For context: some of the missing operations are real gaps, some I feel like don't really make sense in a test environment -- eg `list_bucket_analytics_configurations` etc)

Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: > qq: do we have cases that we test with real s3 bucket? 

We do -- on phone, but at least from memory, I think those live in system tests in `test_object_references.py`.  In my mind moto isn't meant to replace those.

That said, if we for some reason **do** want to use `moto` with a "real" test server, it does have a "server mode" now: https://docs.getmoto.org/en/latest/docs/server_mode.html

fair warning, haven't looked into this feature myself very much.  I've mainly just used moto in the past for mocking/stubbing the various AWS Python APIs

Comment (pingleiwandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: Oh right, for system test we are using real server. In theory we can also use minio, which the local test container already ship with. There might be issue on the virtual host vs path style issue. Also minio is no longer active ....

File: tests/unit_tests/test_artifacts/test_references_s3.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    """An empty, version-enabled bucket; returns its name."""
+    mock_s3.create_bucket(Bucket="my-bucket")
+    mock_s3.put_bucket_versioning(
+        Bucket="my-bucket", VersioningConfiguration={"Status": "Enabled"}
+    )
+    return "my-bucket"
+
+
+def test_add_s3_reference_object(
+    mock_s3, versioned_bucket, mock_s3_content, mock_s3_etag, artifact
```
Comment (jacobromero) [inline tests/unit_tests/test_artifacts/test_references_s3.py]: nit: add a `,` after `artifact` so each parameter will be on its ownline.

File: tests/unit_tests/test_artifacts/test_references_s3.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+):
+    version_id = mock_s3.put_object(
+        Bucket=versioned_bucket, Key="my_object.pb", Body=mock_s3_content
```
Comment (jacobromero) [inline tests/unit_tests/test_artifacts/test_references_s3.py]: same here

File: tests/unit_tests/test_artifacts/test_references_s3.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            "size": 10,
+        },
+    }
+
+
+def test_add_s3_reference_path(
+    capsys, mock_s3, bucket, mock_s3_content, mock_s3_etag, artifact
+):
+    mock_s3.put_object(Bucket=bucket, Key="my_object.pb", Body=mock_s3_content)
+    mock_s3.put_object(Bucket=bucket, Key="my_other_object.pb", Body=mock_s3_content)
```
Comment (jacobromero) [inline tests/unit_tests/test_artifacts/test_references_s3.py]: Would it be more valuable to make the second object live under a directory?
e.g. `mock_s3.put_object(Bucket=bucket, Key="my_dir/my_other_object.pb", Body=mock_s3_content)`

Note: I am assuming that adding a reference would include objects in those sub directories.

Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: Yep strictly speaking this changes the test but let's do this

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py [hunk 5]
```diff
@@ -1239,10 +1018,15 @@ def test_http_storage_handler_uses_etag_for_digest(
...
-    mock_boto(artifact, version_id="")
-    artifact.add_reference("s3://my-bucket/my_object.pb")
+def test_s3_storage_handler_load_path_missing_reference(
+    monkeypatch, user, aws_credentials, artifact
+):
+    # Reference an S3 object that exists when the reference is added.
+    with mock_aws():
+        client = boto3.client("s3", region_name="us-east-1")
+        client.create_bucket(Bucket="my-bucket")
+        client.put_object(Bucket="my-bucket", Key="my_object.pb", Body=b"0123456789")
```
Comment (tonyyli-wandb) [inline tests/system_tests/test_artifacts/test_wandb_artifacts.py]: TODO: pull out an `s3` fixture for the mock S3 client, following similar pattern(s) + names as the new unit test fixtures in this PR.

Also do this for similar snippets in this file (below)

File: tests/unit_tests/test_artifacts/test_references_s3.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    _, err = capsys.readouterr()
+    assert "Generating checksum" in err
+
+
+@mark.xfail(
+    reason=(
+        "max_objects should cap how many objects a directory reference tracks, "
+        "but S3Handler.store_path doesn't enforce it yet: it asks S3 for only "
+        "max_objects objects, so it never sees that more exist and never raises. "
+        "The next PR in this stack fixes the handler, after which this passes."
```
Comment (tonyyli-wandb) [inline tests/unit_tests/test_artifacts/test_references_s3.py]: TODO: condense this explanation for improved brevity and clarity. do not reference "next PR in this stack", since it's unhelpful to other devs -- especially if we decide (for whatever reason) to not ultimately merge the upstack PR

File: tests/unit_tests/test_artifacts/test_references_s3.py (hunk 2 above)
Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: done

File: tests/unit_tests/test_artifacts/test_references_s3.py (hunk 3 above)
Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: done

File: tests/unit_tests/test_artifacts/test_references_s3.py (hunk 4 above)
Comment (tonyyli-wandb) [reply tests/unit_tests/test_artifacts/test_references_s3.py]: done

**Comment by tonyyli-wandb** [comment]
@jacobromero @pingleiwandb  when you get a chance, would you mind taking another look?

For transparency, I've updated the PR so it completely replaces the hand-rolled `mock_boto()` helper with `moto` fixture(s) in system tests, too.  (Previously, there were some system tests that still required mocked S3 operations, but since they also make other calls to the local test server, they couldn't be moved to `unit_tests` with the others)

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py
Comment (jacobromero) [inline tests/system_tests/test_artifacts/test_wandb_artifacts.py]: its outside the scope of this PR, but is it necessary to start another run here?

I just tried and the test passes when I remove that line.

**Review by jacobromero — APPROVED** [review]

File: tests/system_tests/test_artifacts/test_wandb_artifacts.py
Comment (tonyyli-wandb) [reply tests/system_tests/test_artifacts/test_wandb_artifacts.py]: Yeah I admit I really wanted to remove this too TBH.

I don't **think** it's necessary, but seemed safer to leave as-is, in case it just happens to be covering some codepath/edge case (that's not as obvious)

Comment (tonyyli-wandb) [reply tests/system_tests/test_artifacts/test_wandb_artifacts.py]: eh.  AFAICT it's really unnecessary.  Let's drop it

**Comment by tonyyli-wandb** [comment]
### Merge activity

* **Jun 29, 7:54 PM UTC**: @tonyyli-wandb merged this pull request with [Graphite](https://app.graphite.com/github/pr/wandb/wandb/12090).
